sound/soc/sof/sof-audio.c

```diff
@@ -307,23 +307,37 @@ int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, in
 	const struct ipc_tplg_ops *ipc_tplg_ops = sdev->ipc->ops->tplg;
 	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
 	struct snd_soc_dapm_widget *widget;
-	int i, ret, num_widgets;
+	int i, ret, num_widgets = 0;
+	int start, step;
 
 	/* nothing to set up */
 	if (!list)
 		return 0;
 
+	/* set up DAI widget first for capture */
```

Comment:
why? Is this a bug fix or an improvement?

Reply:
This is in preparation for IPC4, which requires the DAI widget to be set up first but there's more to the story than that. But first I want to see if this have any negative impact on IPC3.

Reply:
@ranj063  maybe we could add a `direction` parameter to for_each_dapm_widgets. Or maybe we can create a new macro so that the code becomes easier to understand.

Also, do you know the reason IPC4 asks for capture to be set first?

Reply:
I agree it's not so obvious why reverse iteration guarantees capture DAIs will be first. Would be cleaner to iterate capture widgets first and not rely on the order?

Reply:
I also don't really get the notion of 'direction' for a DAPM graph. Loops are permitted, and if we have a widget that connects a capture and playback dai, how would you define its direction? This concept seems to make sense only for straight-line pipelines.

Reply:
Another example would be a widget that is just before a demux that feeds into both capture and playback, e.g. for echo reference. What would the direction be?

Reply:
@dbaluta @kv2019i @plbossart Im going to clsoe this and come back with a better solution that doesnt involve using ht edirection of list traversal at all